Add TopologicalVector

gtauzin opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the guidelines for contributing: https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines
-->

**Reference issues/PRs**
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.

Pull requests which are not related to open issues may be rejected!
If suggesting a new feature or change, please discuss it in an issue first.
If fixing a bug, there should be an issue describing it with steps to reproduce.
-->

**Types of changes**
<!--
What types of changes does your code introduce? Put an `x` in all the boxes that apply.
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Description**
Add TopologicalVector.

**Screenshots (if appropriate)**

**Any other comments?**

**Checklist**
<!--
Go over all the following points, and put an `x` in all the boxes that apply. 
If you're unsure about any of these, don't hesitate to ask. We're here to help!
-->
- [x] I have read the [guidelines for contributing](https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines).
- [x] My code follows the code style of this project. I used `flake8` to check my Python changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed. I used `pytest` to check this on Python tests.

<!--
We value all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Thanks for contributing!
-->

Commit:
Add TopologicalVector

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@inait.ai>

ulupo committed:
Merge branch 'master' into vector

ulupo committed:
Merge branch 'master' into vector

ulupo committed:
Fix errors and linting, fix self.homology_dimensions_

ulupo committed:
Add DistanceMetric import

ulupo committed:
Add Seel also and References in TopologicalVector

ulupo committed:
Add caveats on properties of X

ulupo committed:
Merge branch 'master' into vector

ulupo committed:
Fix doc entries

ulupo committed:
Fix linting

ulupo committed:
Reorder doc entries

ulupo committed:
Fix See alsos in diagrams/features.py

ulupo commented on gtda/diagrams/features.py:
```diff
@@ -716,3 +716,184 @@ def transform(self, X, y=None):
+        distances = self._distance_function.pairwise(Xd)
+
+        Xd[:, 1] = Xd[:, 1] - Xd[:, 0]
+        min_persistence = 0.5 * np.minimum(Xd[:, 1], Xd[:, 1].T)
```
@gtauzin could you explain this line to me? I expect `Xd[:, 1]` to be a 1d array at this point, and if true this means that `Xd[:, 1]` are `Xd[:, 1].T` are equal arrays.

gtauzin replied:
I believe you are right, this line must have remained for my own experiments.

ulupo replied:
Should I change this line to
```
min_persistence = 0.5 * Xd[:, 1]
```
? I also think this logic should be documented.

ulupo committed:
Fixes in ComplexPolynomial and TopologicalVector

- Allow tuple n_coefficients
- Fix explanation of behaviour when n_coefficients is None
- Make homology_dimensions_ and n_coefficients_ tuples
- Fix shape of Xt in ComplexPolynomial .transform docs
- Allow tuple and list n_distances in TopologicalVector, fix docs
- Simplify docs for metric in TopologicalVector
- Make n_distances_ a tuple
- Add X input check in TopologicalVector.fit
- Begin fixing TopologicalVector.transform logic

ulupo committed:
Improve tests of ComplexPolynomial

ulupo committed:
Continue adapting the logic of TopologicalVector to list n_distances

ulupo committed:
Place validate_params after array checks as in rest of library

VascoSch92 commented on gtda/diagrams/features.py:
```diff
@@ -642,27 +642,28 @@ def fit(self, X, y=None):
-            self.n_coefficients_ = [_homology_dimensions_counts[dim]
-                                    for dim in self.homology_dimensions_]
-        elif type(self.n_coefficients) == list:
+            self.n_coefficients_ = \
```
just write 
```python
self.n_coefficients_ = tuple(
    [ _homology_dimensions_counts[dim] for dim in self.homology_dimensions_]
)
```

VascoSch92 commented on gtda/diagrams/features.py:
```diff
@@ -642,27 +642,28 @@ def fit(self, X, y=None):
+            self.n_coefficients_ = \
+                tuple([_homology_dimensions_counts[dim]
+                       for dim in self.homology_dimensions_])
+        elif type(self.n_coefficients) in (list, tuple):
```
this is not how we check the type.

it should be `elif isistance(self.n_coefficients, (list, tuple))`

You should use the modes Typing instead of the primitive types.

VascoSch92 commented on gtda/diagrams/features.py:
```diff
@@ -642,27 +642,28 @@ def fit(self, X, y=None):
                     )
             self.n_coefficients_ = self.n_coefficients
         else:
             self.n_coefficients_ = \
```
please rewrite it better

VascoSch92 commented on gtda/diagrams/features.py:
```diff
@@ -642,27 +642,28 @@ def fit(self, X, y=None):
-                    f'`n_coefficients` has been passed as a list of length '
-                    f'{len(self.n_coefficients)} while diagrams in `X` have '
-                    f'{_n_homology_dimensions} homology dimensions.'
+                    f'`n_coefficients` has length {len(self.n_coefficients)} '
```
difficult to read this message

VascoSch92 commented on gtda/diagrams/features.py:
```diff
@@ -642,27 +642,28 @@ def fit(self, X, y=None):
-                [self.n_coefficients] * _n_homology_dimensions
+                tuple([self.n_coefficients] * _n_homology_dimensions)
 
         self._polynomial_function = \
```
please rewrite that

VascoSch92 commented on gtda/diagrams/features.py:
```diff
@@ -696,20 +697,215 @@ def transform(self, X, y=None):
+        }
+
+    def __init__(self, n_distances=10, metric='chebyshev', metric_params={},
+                 n_jobs=None):
```
put a comma after `n_jobs=None`.

Why you chose these default parameters?

VascoSch92 commented on gtda/diagrams/features.py:
```diff
@@ -696,20 +697,215 @@ def transform(self, X, y=None):
+                                               counts))
+
+        if self.n_distances is None:
+            self.n_distances_ = \
```
rewrite that please

VascoSch92 commented on gtda/diagrams/features.py:
```diff
@@ -696,20 +697,215 @@ def transform(self, X, y=None):
+                    )
+            self.n_distances_ = self.n_distances
+        else:
+            self.n_distances_ = \
```
rewrite that please

VascoSch92 commented on gtda/diagrams/features.py:
```diff
@@ -696,20 +697,215 @@ def transform(self, X, y=None):
+            self.n_distances_ = \
+                tuple([self.n_distances] * self.homology_dimensions_)
+
+        self._distance_function = \
```
rewrite that please
